subprojects/core/src/main/java/org/gradle/internal/operations/DefaultBuildOperationQueue.java

```diff
@@ -71,8 +71,7 @@ public void add(final T operation) {
             workQueue.add(operation);
             pendingOperations++;
             workAvailable.signalAll();
-            if (workerCount == 0 || workerCount < workerLeases.getMaxWorkerCount() - 1) {
-                // `getMaxWorkerCount() - 1` because main thread executes work as well. See https://github.com/gradle/gradle/issues/3273
+            if (workerCount == 0 || workerCount < workerLeases.getMaxWorkerCount()) {
```

Comment:
I don't think "main thread" here means the main thread of the build.  It means the thread we are submitting work from and calling `waitForCompletion()` (where we also process items from the queue).  I think we'll effectively be processing workerLeases+1 items concurrently now (i.e. one of the children will borrow the a worker lease from the submitting thread, meanwhile the submitting thread will also be doing work on the queue).

We need the submitting thread to also process work because of the problem described in https://github.com/gradle/gradle/issues/3273.

Reply:
Thanks, @ghale! I've reverted that change.